### PR TITLE
chore: skip pointer implementation

### PR DESCRIPTION
**Pull request by savannahar68**
## What does this PR do?
- Summarize your change (**mandatory**)

## How does this PR work? (optional)
- Need a brief introduction for the changed logic (optional)

## Refer to a related PR or issue link
- (Related PR or issue)

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests. (for non-documentation changes)

### CODE REVIEW

File: tsldb/src/log/postings_block_compressed.rs [hunk 1]
```diff
@@ -10,7 +10,7 @@ use crate::utils::error::TsldbError;
 use crate::utils::sync::RwLock;
 
 /// Represents a delta-compressed PostingsBlock.
-#[derive(Debug, Deserialize, Serialize)]
+#[derive(Debug, Deserialize, Serialize, Clone, Copy)]
```
Comment (savannahar68) [inline tsldb/src/log/postings_block_compressed.rs:13]: @vinaykakade what can be parallel implementation of Copy train for PostingsBlockCompressed?

Without copy trait the search function in segment.rs won't work while copying PostingsBlockCompressed in postings_lists

Comment (vinaykakade) [reply tsldb/src/log/postings_block_compressed.rs]: Added a `clone` method to PostingsBlockCompressed. Let us not depend on the implicit `copy` usage (as it gets expensive under the feet quickly), but instead rely on explicitly calling Clone wherever needed.

This has causes some errors in segment.rs. Let me know if you can take a look.

File: tsldb/src/segment_manager/segment.rs [hunk 2]
```diff
@@ -323,32 +332,144 @@ impl Segment {
...
+          acc_index += 1;
+        }
+
+        if acc_index < accumulator.len() && accumulator[acc_index] > initial_values[initial_index] {
+          // If current accumulator element is in between current initial_value and next initial_value
+          // then check the existing posting block for matches with accumlator
+          if initial_index + 1 < initial_values.len()
+            && accumulator[acc_index] < initial_values[initial_index + 1]
+          {
+            let posting_block = PostingsBlock::try_from(&posting_list).unwrap();
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:390]: > PostingsBlock::try_from(&posting_list).unwrap();

Is this intentional? (try_from is on postings_list)

**Comment by savannahar68** [comment]
***Now running Infino via tsldb library***
Infino time required for insertion: 337.44ms
Checking size of directory: index
Infino index size = 1839158 bytes
Infino time required for searching 1 word query is : 1.90ms. Num of results 6784
Infino time required for searching 2 word query is : 5.63µs. Num of results 0
Infino time required for searching 3 word query is : 284.50µs. Num of results 0
Infino time required for searching 4 word query is : 187.00µs. Num of results 134
Infino time required for searching 5 word query is : 191.00µs. Num of results 0
Infino time required for searching 6 word query is : 102.83µs. Num of results 0
Infino time required for searching 7 word query is : 245.92µs. Num of results 0


***Now running Infino via REST API client***
Infino REST time required for insertion: 766.68ms
We haven't yet added flush API to infino client, so can't get accurate index size via API
Infino REST time required for searching 1 word query is 3.64ms
Infino REST time required for searching 2 word query is 2.77ms
Infino REST time required for searching 3 word query is 1.98ms
Infino REST time required for searching 4 word query is 885.79µs
Infino REST time required for searching 5 word query is 14.82ms
Infino REST time required for searching 6 word query is 1.73ms
Infino REST time required for searching 7 word query is 826.67µs

**Comment by savannahar68** [comment]
Need to check why there are 0 results
also implementation for last block is left

File: tsldb/src/log/postings_block_compressed.rs [hunk 3]
```diff
@@ -1,3 +1,5 @@
+use core::num;
```
Comment (vinaykakade) [inline tsldb/src/log/postings_block_compressed.rs:1]: Is this crate needed?

File: tsldb/src/log/postings_list.rs [hunk 4]
```diff
@@ -260,12 +240,19 @@ mod tests {
...
       pl.append(i as u32);
-      expected.push(i as u32);
+    }
+    // Created 2D vec for each posting list
+    for i in 0..num_blocks {
+      let mut temp_vec: Vec<u32> = Vec::new();
+      for j in i * BLOCK_SIZE_FOR_LOG_MESSAGES..(i + 1) * BLOCK_SIZE_FOR_LOG_MESSAGES {
+        temp_vec.push(j as u32);
+      }
+      expected_2_d.push(temp_vec);
```
Comment (vinaykakade) [inline tsldb/src/log/postings_list.rs:255]: Is this used in test? Didn't see an asset on this

File: tsldb/src/segment_manager/segment.rs [hunk 5]
```diff
@@ -301,11 +303,18 @@ impl Segment {
...
   /// Boolean queries are not yet supported.
   pub fn search(&self, query: &str, range_start_time: u64, range_end_time: u64) -> Vec<LogMessage> {
-    // TODO: make the implementation below more performant by using the skip pointers (aka initial values)
-    // and not decompressing every block in every postings list.
+    // TODO: make the implementation below more performant by not decompressing every block in every postings list.
     let query_lowercase = query.to_lowercase();
     let terms = query_lowercase.split_whitespace();
-    let mut postings_lists: Vec<Vec<u32>> = Vec::new();
+
+    // // posting lists will contain <list of posting_list<list of posting_block<list of log_ids>>>
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:310]: Is this and the line below needed? Looks like it isn't used.

File: tsldb/src/segment_manager/segment.rs [hunk 6]
```diff
@@ -301,11 +303,18 @@ impl Segment {
...
+    // let mut posting_lists: Vec<Vec<Vec<u32>>> = Vec::new();
+    // initial_values_list wil contain list of initial_values corresponding to every posting_list
+    let mut initial_values_list: Vec<Vec<u32>> = Vec::new();
+
+    // New code
+    // postings lsit will contain list of PostingBlocksCompressed
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:316]: nit: lsit->list

File: tsldb/src/segment_manager/segment.rs [hunk 7]
```diff
@@ -301,11 +303,18 @@ impl Segment {
...
     let query_lowercase = query.to_lowercase();
     let terms = query_lowercase.split_whitespace();
-    let mut postings_lists: Vec<Vec<u32>> = Vec::new();
+
+    // // posting lists will contain <list of posting_list<list of posting_block<list of log_ids>>>
+    // let mut posting_lists: Vec<Vec<Vec<u32>>> = Vec::new();
+    // initial_values_list wil contain list of initial_values corresponding to every posting_list
+    let mut initial_values_list: Vec<Vec<u32>> = Vec::new();
+
+    // New code
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:315]: nit: redundant comment

File: tsldb/src/segment_manager/segment.rs [hunk 8]
```diff
@@ -323,32 +332,161 @@ impl Segment {
...
         }
       };
-      let flattened_postings_list = postings_list.flatten().clone();
-      postings_lists.push(flattened_postings_list);
+      let inital_values = postings_list.get_initial_values().read().unwrap().clone();
+
+      // posting_lists.push(flatenned_postings_list_2_d);
+      initial_values_list.push(inital_values);
+
+      // New code
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:340]: nit: redundant comment

File: tsldb/src/segment_manager/segment.rs [hunk 9]
```diff
@@ -323,32 +332,161 @@ impl Segment {
...
 
     if postings_lists.is_empty() {
       // No postings list
       return vec![];
     }
 
-    let mut iter = postings_lists.iter();
+    let mut accumulator = Vec::new();
+
+    // Create accumulator from 1st posting list from postings_lists. Which is flatten the 1st Vec<PostingsBlockCompressed> to Vec<u32>
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:363]: Important - Can you create an accumulator from the term that has the shortest postings list? That will quickly make the performance much better.

See the algorithm at end of https://nlp.stanford.edu/IR-book/html/htmledition/processing-boolean-queries-1.html and using 'SortByIncreasingFrequency'

File: tsldb/src/segment_manager/segment.rs [hunk 10]
```diff
@@ -323,32 +332,161 @@ impl Segment {
...
+    // postings_lists.first() will give Vec<PostingsBlockCompressed> which needs to be iterated on to get Vec<u32>
+    let first_posting_blocks = postings_lists.first().unwrap();
+    for posting_block in first_posting_blocks {
+      let posting_block = PostingsBlock::try_from(posting_block).unwrap();
+      accumulator.append(&mut posting_block.get_log_message_ids().read().unwrap().clone());
+    }
+
+    for i in 1..postings_lists.len() {
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:371]: Can you please add a comment on why starting from 1? (The 0th one is the accumulator, but it isn't obvious without reading more code)

File: tsldb/src/segment_manager/segment.rs [hunk 11]
```diff
@@ -323,32 +332,161 @@ impl Segment {
...
+      let initial_values = &initial_values_list[i];
+
+      let mut temp_result_set = Vec::new();
+      let mut acc_index = 0;
+      let mut posting_index = 0;
+      let mut initial_index = 0;
+
+      while acc_index < accumulator.len() && posting_index < posting_list.len() {
+        // If current accumulator element < initial_value element it means that
+        // accumulator values is smaller than what current posting_block will have
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:382]: nit: accumulator values -> accumulator value

**Comment by vinaykakade** [comment]
Thanks @savannahar68.

The PR and the algorithm look good. Before checking in, please do the following:
1. Update the benchmark with larger Apache log dataset, and see that performance esp for search improves compared to current main,
2. Add multiple tests to cover the edge cases. Really needs to be bullet proof as we'll likely work on additional optimizations here.

File: tsldb/src/log/postings_list.rs [hunk 12]
```diff
@@ -86,31 +86,11 @@ impl PostingsList {
     &self.initial_values
   }
 
-  #[cfg(test)]
+  // #[cfg(test)]
```
Comment (vinaykakade) [inline tsldb/src/log/postings_list.rs:89]: Best to delete this line altogether instead of commenting it out

File: tsldb/src/log/postings_list.rs [hunk 13]
```diff
@@ -295,8 +273,5 @@ mod tests {
       initial_values.read().unwrap()[num_blocks - 1],
       ((num_blocks - 1) * BLOCK_SIZE_FOR_LOG_MESSAGES) as u32
     );
-
-    let received = pl.flatten();
-    assert_eq!(expected, received);
```
Comment (vinaykakade) [inline tsldb/src/log/postings_list.rs:300]: We are losing an existing test here. Can we keep flatten to be with `#[cfg(test)]` and use it here?

File: tsldb/src/log/postings_block_compressed.rs [hunk 14]
```diff
@@ -116,6 +124,20 @@ impl TryFrom<&PostingsBlock> for PostingsBlockCompressed {
   }
 }
 
+impl Clone for PostingsBlockCompressed {
```
Comment (vinaykakade) [inline tsldb/src/log/postings_block_compressed.rs:127]: Will be great to add a test for the clone. And perhaps get this change in before other changes.

File: tsldb/src/segment_manager/segment.rs [hunk 15]
```diff
@@ -323,32 +333,182 @@ impl Segment {
...
-      let flattened_postings_list = postings_list.flatten().clone();
-      postings_lists.push(flattened_postings_list);
+      let inital_values = postings_list.get_initial_values().read().unwrap().clone();
+
+      // posting_lists.push(flatenned_postings_list_2_d);
+      initial_values_list.push(inital_values);
+
+      // Extract List of PostingBlockCompressed from posting list
+      // posting_block_compressed is Vec<PostingsBlockCompressed> which is extracted by cloning get_postings_list_compressed from postings_lists
+      let mut posting_block_compressed: Vec<PostingsBlockCompressed> = Vec::new();
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:343]: Can we rename this variable to postings_block_compressed_vec? That way, it is clearer that it is a vector while reading rest of the code

File: tsldb/src/segment_manager/segment.rs [hunk 16]
```diff
@@ -323,32 +333,182 @@ impl Segment {
...
+              break;
+            }
+          }
+        }
 
-    // Start with the first postings list as the initial result
-    let mut log_message_ids = iter.next().unwrap().clone();
+        initial_index += 1;
+        posting_index += 1;
+        posting_index += 1;
```
Comment (vinaykakade) [inline tsldb/src/segment_manager/segment.rs:498]: twice? looks like a bug

**Comment by vinaykakade** [comment]
Thanks @savannahar68. Added comments. In addition to items mentioned in https://github.com/infinohq/infino/pull/82#issuecomment-1750070916, one more TODO below:

3. Check if you don't need to compress the last block. It may impact the performance negatively.

File: tsldb/src/log/postings_list.rs (hunk 12 above)
Comment (savannahar68) [reply tsldb/src/log/postings_list.rs]: this is resolved

File: tsldb/src/segment_manager/segment.rs (hunk 15 above)
Comment (savannahar68) [reply tsldb/src/segment_manager/segment.rs]: addressed

File: tsldb/src/segment_manager/segment.rs (hunk 16 above)
Comment (savannahar68) [reply tsldb/src/segment_manager/segment.rs]: fixed this

File: tsldb/src/log/postings_list.rs (hunk 13 above)
Comment (savannahar68) [reply tsldb/src/log/postings_list.rs]: added the test back

File: tsldb/src/log/postings_block_compressed.rs (hunk 14 above)
Comment (savannahar68) [reply tsldb/src/log/postings_block_compressed.rs:127]: added tests

File: benches/output.txt [hunk 17]
```diff
@@ -1,107 +1,102 @@
...
+PrometheusClient time required for searching 9.35ms
+PrometheusClient time required for searching 4.33ms
+PrometheusClient time required for searching 2.28ms
+PrometheusClient time required for searching 1.36ms
+PrometheusClient time required for searching 1.18ms
+PrometheusClient time required for searching 1.57ms
+PrometheusClient time required for searching 2.46ms
+PrometheusClient time required for searching 2.89ms
+PrometheusClient time required for searching 1.36ms
+Prometheus timeseries search avg 2678 nanoseconds
```
Comment (vinaykakade) [inline benches/output.txt:94]: The unit here won't be nanoseconds?

File: benches/README.md [hunk 18]
```diff
@@ -48,48 +50,42 @@ $ cd benches
...
 ### Timeseries search latency
 
 Average over 10 queries on time series.
 
-| Data points | Prometheus | Infino |
-| ----------- | ---------- | ---------- |
-| Search Latency | 1796 microseconds | 602 microseconds |
+| Data points    | Prometheus        | Infino            |
+| -------------- | ----------------- | ----------------- |
+| Search Latency | 2678 microseconds | 2864 microseconds |
```
Comment (vinaykakade) [inline benches/README.md:91]: Please check the units / see whether this is being computed correctly. (The output.txt is inconsistent with these numbers.)

Comment (savannahar68) [reply benches/README.md]: fixed

File: benches/output.txt (hunk 17 above)
Comment (savannahar68) [reply benches/output.txt:94]: it is ms converted to ns, so looks right to me
